mlir/lib/Mitigation/Transforms/MitigationMethods/Zne.cpp

```diff
@@ -36,31 +37,131 @@ namespace mitigation {
 
 LogicalResult ZneLowering::match(mitigation::ZneOp op) const { return success(); }
 
+// TODO: Optimize the traversal of call graphs (currently used twice)
+// Also all functions exploree in the call graph get their ZNE version.
 void ZneLowering::rewrite(mitigation::ZneOp op, PatternRewriter &rewriter) const
 {
+    auto point = rewriter.saveInsertionPoint();
     Location loc = op.getLoc();
-
+    auto moduleOp = op->getParentOfType<ModuleOp>();
     // Number of folds
     auto numFolds = op.getNumFolds();
     RankedTensorType numFoldType = cast<RankedTensorType>(numFolds.getType());
     const auto sizeInt = numFoldType.getDimSize(0);
 
     // Folding type
     auto foldingAlgorithm = op.getFolding();
-
-    // Create the folded circuit function
-    FlatSymbolRefAttr foldedCircuitRefAttr =
-        getOrInsertFoldedCircuit(loc, rewriter, op, foldingAlgorithm);
-    func::FuncOp foldedCircuit =
-        SymbolTable::lookupNearestSymbolFrom<func::FuncOp>(op, foldedCircuitRefAttr);
-
+    auto calleeOp = SymbolTable::lookupNearestSymbolFrom<func::FuncOp>(op, op.getCalleeAttr());
+
+    func::FuncOp fnFoldedOp;
+
+    // Traverse the callgraph and replace qnodes with folded qnodes.
+    if (!calleeOp->hasAttr("qnode")) {
```

Comment:
This block would definitely benefit from outlining different pieces into helper functions 😅

Reply:
Added one function for checking that it contains qnodes and one for creating the .zne counterpart function.